sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -546,6 +546,10 @@ case class DataSource(
       case dataSource: CreatableRelationProvider =>
         SaveIntoDataSourceCommand(data, dataSource, caseInsensitiveOptions, mode)
       case format: FileFormat =>
+        if (DataSource.isBuiltInFileBasedDataSource(format) && data.schema.size == 0) {
```

Comment:
We don't need this check. `FileFormat` is internal, we don't need to distinguish between built-in file format or external ones.

Reply:
actually we just need [this check](https://github.com/apache/spark/pull/20579/files#diff-ee26d4c4be21e92e92a02e9f16dbc285R71) here.